lib/astarte_data_access/entities/realm.ex

```diff
@@ -0,0 +1,912 @@
+    query = """
+    CREATE TABLE #{keyspace_name}.simple_triggers (
+      object_id uuid,
+      object_type int,
+      parent_trigger_id uuid,
+      simple_trigger_id uuid,
+      trigger_data blob,
+      trigger_target blob,
+
+      PRIMARY KEY ((object_id, object_type), parent_trigger_id, simple_trigger_id)
+    )
+    """
+
+    with {:ok, %Xandra.SchemaChange{}} <- CSystem.execute_schema_change(conn, query) do
+      :ok
+    end
+  end
+
+  defp create_grouped_devices_table(conn, realm_name) do
+    query = """
+    CREATE TABLE #{realm_name}.grouped_devices (
+      group_name varchar,
+      insertion_uuid timeuuid,
+      device_id uuid,
+      PRIMARY KEY (
+        (group_name), insertion_uuid, device_id
+      )
+    )
+    """
+
+    with {:ok, %Xandra.SchemaChange{}} <- CSystem.execute_schema_change(conn, query) do
+      :ok
+    end
+  end
+
+  defp create_deletion_in_progress_table(conn, realm_name) do
+    query = """
+    CREATE TABLE #{realm_name}.deletion_in_progress (
+      device_id uuid,
+      vmq_ack boolean,
+      dup_start_ack boolean,
+      dup_end_ack boolean,
+      PRIMARY KEY (device_id)
+    )
+    """
+
+    with {:ok, %Xandra.SchemaChange{}} <- CSystem.execute_schema_change(conn, query) do
+      :ok
+    end
+  end
+
+  defp insert_realm_public_key(conn, realm_name, public_key_pem) do
+    query = """
+    INSERT INTO #{realm_name}.kv_store (
+      group,
+      key,
+      value
+    )
+    VALUES (
+      'auth',
+      'jwt_public_key_pem',
+      varcharAsBlob(:public_key_pem)
+    )
+    """
+
+    params = %{
+      public_key_pem: public_key_pem
+    }
+
+    with {:ok, _} <- XandraUtils.execute_query(conn, query, params, consistency: :each_quorum) do
+      :ok
+    end
+  end
+
+  defp insert_realm_astarte_schema_version(conn, keyspace_name, realm_schema_version) do
+    query = """
+    INSERT INTO #{keyspace_name}.kv_store (
+      group,
+      key,
+      value
+    )
+    VALUES (
+      'astarte',
+      'schema_version',
+      bigintAsBlob(:realm_schema_version)
+    )
+    """
+
+    params = %{
+      realm_schema_version: realm_schema_version
+    }
+
+    with {:ok, _} <- XandraUtils.execute_query(conn, query, params, consistency: :each_quorum) do
+      :ok
+    end
+  end
+
+  defp remove_realm(conn, astarte_keyspace_name, realm_name) do
+    query = """
+    DELETE FROM
+      #{astarte_keyspace_name}.realms
+    WHERE
+      realm_name = :realm_name
+    """
+
+    params = %{
+      realm_name: realm_name
+    }
+
+    with {:ok, _} <- XandraUtils.execute_query(conn, query, params, consistency: :each_quorum) do
+      :ok
+    end
+  end
+
+  defp insert_realm(conn, astarte_keyspace_name, realm_name, device_registration_limit) do
+    query = """
+    INSERT INTO #{astarte_keyspace_name}.realms (
+      realm_name,
+      device_registration_limit
+    )
+    VALUES (
+      :realm_name,
+      :device_registration_limit
+    )
+    """
+
```

Comment:
defaulting to `nil` it becomes useless

Reply:
We should make sure that the check @eddbbt outlined is still in place, either here or in https://github.com/astarte-platform/astarte/pull/999. It seems it's not, but maybe I lost the detail in between the two PRs @shinnokdisengir

Reply:
@eddbbt that row was a bug. I've opened this issue https://github.com/astarte-platform/astarte_data_access/issues/87, my solution could resolve

Reply:
@shinnokdisengir I think that's an intended behaviour as that feature was created to set an upper limit, not to stop devices from connecting completely. @Annopaolo would you like to check what was the original scope of this check?

Also, even not it's not the reason behind, BUT completely blocking devices from connecting (but not disabling current connected devices) could be used as a go-to scenario in case of realm deletion with the marketplace version of astarte, effectively providing a grace period. Better think about that